### PR TITLE
Skip the structured dataset format check when the expected format is empty

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
For more detail, check https://github.com/flyteorg/flyte/issues/2864

Failed to run BQ task when the cache is enabled because [type validation](https://github.com/flyteorg/flytepropeller/blob/9cf5368793e5d60113cb819a6e496a7b6accd619/pkg/compiler/validators/typing.go#L157) is failing.

In some cases, the expected type is empty but the input type has value, so skip the format check when the expected format is empty

When the expected format of SD is empty, the dataframe will be converted to a specific format determined by the SD transformer.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
^^^

## Tracking Issue
^^^

## Follow-up issue
_NA_

### CODE REVIEW

**Comment by wild-endeavor** [comment]
@hamersaw are you okay with this now?

File: pkg/compiler/validators/typing.go [hunk 1]
```diff
@@ -154,7 +152,8 @@ func (t structuredDatasetChecker) CastsFrom(upstreamType *flyte.LiteralType) boo
 		}
 		return structuredDatasetCastFromSchema(schemaType, t.literalType.GetStructuredDatasetType())
 	}
-	if !strings.EqualFold(structuredDatasetType.Format, t.literalType.GetStructuredDatasetType().Format) {
+	// Skip the format check here when format is empty. https://github.com/flyteorg/flyte/issues/2864
+	if len(structuredDatasetType.Format) != 0 && len(t.literalType.GetStructuredDatasetType().Format) != 0 && !strings.EqualFold(structuredDatasetType.Format, t.literalType.GetStructuredDatasetType().Format) {
```
Comment (EngHabu) [inline pkg/compiler/validators/typing.go:156]: Should this check move into the structedDatasetCastsFromStruxtedDataset function?

**Comment by hamersaw** [comment]
> @hamersaw are you okay with this now?

One final question cc @wild-endeavor @pingsutw  - what are the valid format mappings for StructuredDatasets? For example, if we expect a type of `SD["csv"]` and we provide a `SD["parquet"]` is that valid in flytekit? Or is an empty format the only thing that maps to every output format? Can someone briefly give a mapping?
```
expect |   valid
------------------------------------
csv    |   empty, csv, parquet?
...
```

**Comment by pingsutw** [comment]
> if we expect a type of SD["csv"] and we provide a SD["parquet"] is that valid in flytekit

No, it's invalid

> is an empty format the only thing that maps to every output format

right.

expect |   valid |
--------| -------|
csv    |   csv, "" |
parquet | parquet, "" |
"" | parquet, "", csv |

**Comment by hamersaw** [comment]
> 

@pingsutw - perfect, thanks so much for explaining. once the checks are working i'm happy to approve!

**Review by hamersaw — APPROVED** [review]
